sm/download/src/download.rs

```diff
@@ -0,0 +1,142 @@
+use crate::error::DownloadError;
+use async_trait::async_trait;
+use backoff::{backoff::Backoff, future::retry, ExponentialBackoff};
+use c8y_smartrest::smartrest_deserializer::SmartRestJwtResponse;
+use mqtt_client::{Client, MqttClient, Topic};
+use reqwest::{self, Url};
+use std::{
+    path::{Path, PathBuf},
+    time::Duration,
+};
+use tedge_utils;
+
+#[async_trait]
+pub trait Downloader {
+    async fn download(&self);
+
+    async fn cleanup_downloaded(&self);
+}
+
+pub async fn download(
+    url: &UrlType,
+    target_dir_path: impl AsRef<Path>,
+    target_file_name: impl AsRef<Path>,
+) -> Result<PathBuf, DownloadError> {
+    // TODO: Validate the url belongs to the tenant and we can use jwt token such that we don't leak credentials
+
+    // Default retry is an exponential retry with a limit of 15 minutes total
+    let response = retry(ExponentialBackoff::default(), || async {
+        Ok(url.clone().get_from_url().await?)
+    })
+    .await?;
+
+    dbg!(&response);
+    let content = response.bytes().await?;
+
+    let temp_path = PathBuf::new().join(&target_dir_path).join("dl.tmp");
+    let target_path = PathBuf::new().join(target_dir_path).join(target_file_name);
+
+    dbg!(&target_path);
+    // Cleanup after `disc full` will happen inside atomic write function.
+    // TODO: Add cleanup on file exists
+    let () = tedge_utils::fs::atomically_write_file_async(
+        temp_path.as_path(),
+        target_path.as_path(),
+        content.as_ref(),
+    )
+    .await?;
+
+    dbg!();
+    Ok(target_path)
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum UrlType {
+    C8y(Url),
+    NonC8y(Url),
+    Unsupported(String),
+}
```

Comment:
This has to be reworked in order to remove any cloud specific code or case.

Something along the lines:

```
struct DownloadUrl {
     url: Url,
     credentials: Credentials 
}

enum Credentials  {
      NoAuth,                   /// any public package repo
      Bearer(String),          /// notably used by C8Y
      Basic(String),            /// useless for now but making sense
}
```